bin/i18n/utils/crowdin_client.rb

```diff
@@ -0,0 +1,221 @@
+        crowdin_dir_path = crowdin_source_path(dir_path)
+        return if crowdin_dir_path.empty?
+
+        crowdin_dir_name = crowdin_source_name(crowdin_dir_path)
+        return if crowdin_dir_name.empty?
+
+        request_offset = 0
+        crowdin_directory = nil
+
+        loop do
+          crowdin_dirs = request(:list_directories, filter: crowdin_dir_name, offset: request_offset, limit: MAX_ITEMS_COUNT)['data']
+
+          crowdin_directory = crowdin_dirs.find {|crowdin_dir| crowdin_dir.dig('data', 'path') == crowdin_dir_path}
+
+          crowdin_directory || crowdin_dirs.size < MAX_ITEMS_COUNT ? break : request_offset += MAX_ITEMS_COUNT
+        end
+
+        crowdin_directory&.dig('data')
+      end
+
+      # Retrieves the Crowdin source file data of the given i18n source file path
+      # https://developer.crowdin.com/api/v2/#operation/api.projects.files.getMany
+      #
+      # @param file_path [String] the i18n source file path
+      # @return [Hash, nil] the Crowdin source file data
+      def get_source_file(file_path)
+        crowdin_file_name = crowdin_source_name(file_path)
+        crowdin_file_path = crowdin_source_path(file_path)
+
+        request_offset = 0
+        crowdin_file = nil
+
+        loop do
+          crowdin_files = request(:list_files, filter: crowdin_file_name, offset: request_offset, limit: MAX_ITEMS_COUNT)['data']
+
+          crowdin_file = crowdin_files.find {|file| file.dig('data', 'path') == crowdin_file_path}
+
+          crowdin_file || crowdin_files.size < MAX_ITEMS_COUNT ? break : request_offset += MAX_ITEMS_COUNT
+        end
+
+        crowdin_file&.dig('data')
+      end
+
+      # Uploads file to Crowdin Storage to obtain a unique `storageId` for adding it to the project later
+      # https://developer.crowdin.com/api/v2/#operation/api.storages.post
+      #
+      # @param file_path [String] the i18n file path
+      # @return [Hash] the Crowdin storage data
+      def add_storage(file_path)
+        request(:add_storage, file_path)['data']
+      end
+
+      # Uploads the given i18n source file to Crowdin project
+      # https://developer.crowdin.com/api/v2/#section/Introduction/File-Upload
+      #
+      # https://developer.crowdin.com/api/v2/#operation/api.projects.files.post
+      # https://developer.crowdin.com/api/v2/#operation/api.projects.files.put
+      #
+      # @param file_path [String] the i18n source file path
+      # @return [Hash] the Crowdin source file data
+      def upload_source_file(file_path)
+        crowdin_storage = add_storage(file_path)
+        crowdin_storage_id = crowdin_storage['id']
+
+        crowdin_file = get_source_file(file_path)
+
+        if crowdin_file
+          request(
+            :update_or_restore_file,
+            crowdin_file['id'],
+            storageId: crowdin_storage_id,
+          )['data']
+        else
+          crowdin_directory = mutex.synchronize {source_directory File.dirname(file_path)}
```

Comment:
I'd love to see a comment explaining why we need a mutex here!

Reply:
I replaced mutex with error handling as we agreed